// ionic2-pincode-input/esm2015/ionic2-pincode-input.js
import { Component, ElementRef, ViewEncapsulation, trigger, state, animate, transition, style, Renderer2, Injectable, NgModule } from '@angular/core';
import { GestureController, BLOCK_ALL, ViewController, Platform, NavParams, Config, App, IonicModule } from 'ionic-angular';
import { assert, isNumber, isPresent, isFunction } from 'ionic-angular/util/util';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class PincodeCmp {
    /**
     * @param {?} _viewCtrl
     * @param {?} _elementRef
     * @param {?} config
     * @param {?} gestureCtrl
     * @param {?} params
     * @param {?} _renderer
     * @param {?} _plt
     */
    constructor(_viewCtrl, _elementRef, config, gestureCtrl, params, _renderer, _plt) {
        this._viewCtrl = _viewCtrl;
        this._elementRef = _elementRef;
        this._renderer = _renderer;
        this._plt = _plt;
        this.stateExpression = 'off';
        this.codeArr = [];
        this.maxLen = 6;
        this.gestureBlocker = gestureCtrl.createBlocker(BLOCK_ALL);
        this.d = params.data;
        this.d.passSize && (this.maxLen = this.d.passSize);
        this.codeArr = new Array(this.maxLen);
        this.codeArr.fill(null);
        this.mode = config.get('mode');
        _renderer.addClass(_elementRef.nativeElement, `pincode-${this.mode}`);
        if (this.d.cssClass) {
            this.d.cssClass.split(' ').forEach(cssClass => {
                if (cssClass.trim() !== '')
                    _renderer.addClass(_elementRef.nativeElement, cssClass);
            });
        }
    }
    /**
     * @return {?}
     */
    ionViewDidLoad() {
        this.stateExpression = 'on';
    }
    /**
     * @return {?}
     */
    ionViewWillEnter() {
        this.gestureBlocker.block();
    }
    /**
     * @return {?}
     */
    ionViewDidLeave() {
        this._plt.focusOutActiveElement();
        this.gestureBlocker.unblock();
    }
    /**
     * @return {?}
     */
    ionViewWillLeave() {
        this._plt.focusOutActiveElement();
    }
    /**
     * @return {?}
     */
    ionViewDidEnter() {
        this._plt.focusOutActiveElement();
        this.enabled = true;
    }
    /**
     * @param {?} num
     * @return {?}
     */
    numClick(num) {
        if (num < 0 || num > 9)
            return;
        const /** @type {?} */ emptyIndex = this.codeArr.indexOf(null);
        if (emptyIndex < this.maxLen - 1) {
            this.codeArr[emptyIndex] = num;
        }
        else if (emptyIndex === this.maxLen - 1) {
            this.codeArr[emptyIndex] = num;
            // If we have a completed PIN handler,
            // call it with the completed PIN.
            if (this.d.pinHandler) {
                this.d.pinHandler(this.getValues())
                    .then(() => {
                    // PIN is valid.
                    this.dismiss('done');
                })
                    .catch(() => {
                    // PIN is invalid.
                    this.restoreClick();
                });
            }
            else {
                this.dismiss('done');
            }
        }
    }
    /**
     * @return {?}
     */
    delClick() {
        let /** @type {?} */ i = this.codeArr.length - 1;
        for (i; i > -1; i--) {
            if (this.codeArr[i] !== null) {
                this.codeArr[i] = null;
                break;
            }
        }
    }
    /**
     * @return {?}
     */
    restoreClick() {
        this.codeArr = new Array(this.maxLen);
        this.codeArr.fill(null);
    }
    /**
     * @param {?} button
     * @return {?}
     */
    btnClick(button) {
        if (!this.enabled) {
            return;
        }
        // keep the time of the most recent button click
        this.lastClick = Date.now();
        let /** @type {?} */ shouldDismiss = true;
        if (button.handler) {
            // a handler has been provided, execute it
            // pass the handler the values from the inputs
            if (button.handler(this.getValues()) === false) {
                // if the return value of the handler is false then do not dismiss
                shouldDismiss = false;
            }
        }
        if (shouldDismiss) {
            this.dismiss(button.role).catch(() => {
                console.debug('alert can not be dismissed');
            });
        }
    }
    /**
     * @param {?} num
     * @return {?}
     */
    isNum(num) {
        if (isNumber(num)) {
            return num <= 9 || num >= 0;
        }
        else {
            return false;
        }
    }
    /**
     * @return {?}
     */
    bdClick() {
        if (this.enabled && this.d.enableBackdropDismiss) {
            this.dismiss('backdrop');
        }
    }
    /**
     * @return {?}
     */
    cancelClick() {
        this.dismiss('cancel');
    }
    /**
     * @return {?}
     */
    forgotClick() {
        this.dismiss('forgot');
    }
    /**
     * @param {?} role
     * @return {?}
     */
    dismiss(role) {
        this.stateExpression = 'off';
        const /** @type {?} */ opts = {
            minClickBlockDuration: 400
        };
        return this._viewCtrl.dismiss(this.getValues(), role, opts);
    }
    /**
     * @return {?}
     */
    getValues() {
        return this.d.encoded(this.codeArr.join(''));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        assert(this.gestureBlocker.blocked === false, 'gesture blocker must be already unblocked');
        this.gestureBlocker.destroy();
    }
}
PincodeCmp.decorators = [
    { type: Component, args: [{
                selector: 'ion-pincode',
                animations: [
                    trigger('openClose', [
                        state('off, void', style({ bottom: '-295px' })),
                        state('on', style({ bottom: '0' })),
                        transition('on <=> off', [animate("400ms cubic-bezier(.36,.66,.04,1)")])
                    ])
                ],
                template: `
        <ion-backdrop (click)="bdClick()" [class.backdrop-no-tappable]="!d.enableBackdropDismiss"></ion-backdrop>
        <div class="pincode-wrapper"
             [class.hide-toolbar]="d.hideToolbar"
             [@openClose]="stateExpression">
            <div class="pincode-toolbar" *ngIf="!d.hideToolbar">
                <button *ngIf="!d.hideCancelButton" ion-button clear (click)="cancelClick()">{{d.cancelButtonText}}
                </button>
                <div class="pincode-title">{{d.title}}</div>
                <button *ngIf="!d.hideForgotPassword" ion-button clear small class="r-btn" (click)="forgotClick()">
                    {{d.forgotPasswordText}}
                </button>
            </div>
            <div class="pincode-input">
                <ion-grid class="pincode-input-grid">
                    <ion-row class="pincode-input-row">
                        <ion-col *ngFor="let item of codeArr; let i = index">
                            <span *ngIf="!d.visibility || (!isNum(codeArr[i]))" class="obscured" [class.filling]="isNum(codeArr[i])"></span>
                            <span *ngIf="d.visibility" class="code">{{item}}</span>
                        </ion-col>
                    </ion-row>
                </ion-grid>
            </div>
            <div class="pincode-button-wrapper">
                <ion-grid>
                    <ion-row>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(1)">1</button>
                        </ion-col>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(2)">2</button>
                        </ion-col>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(3)">3</button>
                        </ion-col>
                    </ion-row>
                    <ion-row>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(4)">4</button>
                        </ion-col>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(5)">5</button>
                        </ion-col>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(6)">6</button>
                        </ion-col>
                    </ion-row>
                    <ion-row>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(7)">7</button>
                        </ion-col>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(8)">8</button>
                        </ion-col>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(9)">9</button>
                        </ion-col>
                    </ion-row>
                    <ion-row>
                        <ion-col>
                            <button ion-button color="light" icon-only (click)="restoreClick()">
                                <ion-icon ios="ios-refresh" name="md-refresh"></ion-icon>
                            </button>
                        </ion-col>
                        <ion-col>
                            <button ion-button color="light" (click)="numClick(0)">0</button>
                        </ion-col>
                        <ion-col>
                            <button ion-button color="light" icon-only (click)="delClick()">
                                <ion-icon ios="ios-backspace" name="md-backspace"></ion-icon>
                            </button>
                        </ion-col>
                    </ion-row>
                </ion-grid>
            </div>
        </div>
    `,
                styles: [`ion-pincode .pincode-wrapper{z-index:10;position:absolute;bottom:0;left:0;right:0;display:block;overflow:hidden;width:100%;max-width:500px;height:295px;margin:auto;background-color:#fcfcfc;border-top:1px solid #c8c7cc;contain:content}ion-pincode .pincode-wrapper.hide-toolbar{height:251px}ion-pincode .pincode-wrapper .button-md{height:4.6rem;font-size:1.8rem;color:#212121}ion-pincode .pincode-md .pincode-wrapper,ion-pincode .pincode-wp .pincode-wrapper{background-color:#fff}ion-pincode .pincode-toolbar{z-index:1;width:100%;contain:strict;display:block;height:44px;border-bottom:.55px solid #c8c7cc;background:#fff}ion-pincode .pincode-toolbar button{margin:0;height:44px;background:0 0;z-index:1}ion-pincode .pincode-toolbar .r-btn{float:right;text-align:left;font-size:12px;text-transform:none;font-weight:400}ion-pincode .pincode-toolbar .pincode-title{text-align:center;width:100%;line-height:44px;position:absolute;left:0;right:0;top:0;bottom:0}ion-pincode .pincode-md .pincode-toolbar,ion-pincode .pincode-wp .pincode-toolbar{border-bottom:none}ion-pincode .pincode-input{margin:16px}ion-pincode .pincode-input .pincode-input-grid{border:1px solid #e4e4e4}ion-pincode .pincode-input .pincode-input-row{text-align:center}ion-pincode .pincode-input .pincode-input-row ion-col{border-left:1px solid #e4e4e4;height:32px;line-height:32px;padding:0;text-align:center}ion-pincode .pincode-input .pincode-input-row ion-col:nth-child(1){border-left:0}ion-pincode .pincode-input .pincode-input-row ion-col span.obscured{width:1em;height:1em;border-radius:1em;display:inline-block;background-color:#eee}ion-pincode .pincode-input .pincode-input-row ion-col span.obscured.filling{background-color:#333}ion-pincode .pincode-input .pincode-input-row ion-col span.code{font-size:1.2em}ion-pincode .pincode-button-wrapper{position:absolute;width:100%;bottom:0}ion-pincode .pincode-button-wrapper .grid,ion-pincode .pincode-button-wrapper ion-col{padding:0}ion-pincode .pincode-button-wrapper ion-col button{border:none;margin:0;width:100%;background-color:#fff;color:#000;-webkit-box-shadow:none;box-shadow:none;border-radius:0}`],
                host: {
                    'role': 'dialog',
                    '[attr.aria-labelledby]': 'hdrId',
                    '[attr.aria-describedby]': 'descId'
                },
                encapsulation: ViewEncapsulation.None,
            },] },
];
/** @nocollapse */
PincodeCmp.ctorParameters = () => [
    { type: ViewController, },
    { type: ElementRef, },
    { type: Config, },
    { type: GestureController, },
    { type: NavParams, },
    { type: Renderer2, },
    { type: Platform, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class PinCode extends ViewController {
    /**
     * @param {?} app
     * @param {?=} opts
     */
    constructor(app, opts = {}) {
        opts.enableBackdropDismiss = isPresent(opts.enableBackdropDismiss) ?
            !!opts.enableBackdropDismiss : true;
        opts.hideForgotPassword = isPresent(opts.hideForgotPassword) ?
            !!opts.hideForgotPassword : false;
        opts.hideCancelButton = isPresent(opts.hideCancelButton) ?
            !!opts.hideCancelButton : false;
        opts.encoded = isFunction(opts.encoded) ? opts.encoded : (c) => {
            return c;
        };
        opts.title = opts.title || 'Password';
        opts.cancelButtonText = opts.cancelButtonText || 'cancel';
        opts.forgotPasswordText = opts.forgotPasswordText || 'forgot password';
        opts.passSize = opts.passSize || 6;
        super(PincodeCmp, opts, null);
        this._app = app;
        //noinspection TypeScriptUnresolvedVariable
        this.isOverlay = true;
    }
    /**
     * @param {?} direction
     * @return {?}
     */
    getTransitionName(direction) {
        let /** @type {?} */ key = (direction === 'back' ? 'alertLeave' : 'alertEnter');
        //noinspection TypeScriptUnresolvedVariable
        return this._nav && this._nav.config.get(key);
    }
    /**
     * @param {?} title
     * @return {?}
     */
    setTitle(title) {
        //noinspection TypeScriptUnresolvedVariable
        this.data.title = title;
    }
    /**
     * @param {?} cssClass
     * @return {?}
     */
    setCssClass(cssClass) {
        //noinspection TypeScriptUnresolvedVariable
        this.data.cssClass = cssClass;
    }
    /**
     * @param {?=} navOptions
     * @return {?}
     */
    present(navOptions = {}) {
        navOptions.minClickBlockDuration = navOptions.minClickBlockDuration || 400;
        return this._app.present(this, navOptions);
    }
}
class PincodeController {
    /**
     * @param {?} _app
     */
    constructor(_app) {
        this._app = _app;
    }
    /**
     * @param {?=} opts
     * @return {?}
     */
    create(opts = {}) {
        return new PinCode(this._app, opts);
    }
}
PincodeController.decorators = [
    { type: Injectable },
];
/** @nocollapse */
PincodeController.ctorParameters = () => [
    { type: App, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class PincodeInputModule {
}
PincodeInputModule.decorators = [
    { type: NgModule, args: [{
                imports: [IonicModule, CommonModule],
                declarations: [PincodeCmp],
                providers: [PincodeController],
                entryComponents: [PincodeCmp],
            },] },
];
/** @nocollapse */
PincodeInputModule.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Created by hsuanlee on 2017/3/19.
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { PinCode, PincodeController, PincodeCmp, PincodeInputModule };
