// ionic2-pincode-input/bundles/ionic2-pincode-input.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('ionic-angular'), require('ionic-angular/util/util'), require('@angular/common')) :
	typeof define === 'function' && define.amd ? define('ionic2-pincode-input', ['exports', '@angular/core', 'ionic-angular', 'ionic-angular/util/util', '@angular/common'], factory) :
	(factory((global['ionic2-pincode-input'] = {}),global.ng.core,global.ionicAngular,global.util,global.ng.common));
}(this, (function (exports,core,ionicAngular,util,common) { 'use strict';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0
THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.
See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */
var extendStatics = Object.setPrototypeOf ||
    ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
    function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var PincodeCmp = /** @class */ (function () {
    function PincodeCmp(_viewCtrl, _elementRef, config, gestureCtrl, params, _renderer, _plt) {
        this._viewCtrl = _viewCtrl;
        this._elementRef = _elementRef;
        this._renderer = _renderer;
        this._plt = _plt;
        this.stateExpression = 'off';
        this.codeArr = [];
        this.maxLen = 6;
        this.gestureBlocker = gestureCtrl.createBlocker(ionicAngular.BLOCK_ALL);
        this.d = params.data;
        this.d.passSize && (this.maxLen = this.d.passSize);
        this.codeArr = new Array(this.maxLen);
        this.codeArr.fill(null);
        this.mode = config.get('mode');
        _renderer.addClass(_elementRef.nativeElement, "pincode-" + this.mode);
        if (this.d.cssClass) {
            this.d.cssClass.split(' ').forEach(function (cssClass) {
                if (cssClass.trim() !== '')
                    _renderer.addClass(_elementRef.nativeElement, cssClass);
            });
        }
    }
    PincodeCmp.prototype.ionViewDidLoad = function () {
        this.stateExpression = 'on';
    };
    PincodeCmp.prototype.ionViewWillEnter = function () {
        this.gestureBlocker.block();
    };
    PincodeCmp.prototype.ionViewDidLeave = function () {
        this._plt.focusOutActiveElement();
        this.gestureBlocker.unblock();
    };
    PincodeCmp.prototype.ionViewWillLeave = function () {
        this._plt.focusOutActiveElement();
    };
    PincodeCmp.prototype.ionViewDidEnter = function () {
        this._plt.focusOutActiveElement();
        this.enabled = true;
    };
    PincodeCmp.prototype.numClick = function (num) {
        var _this = this;
        if (num < 0 || num > 9)
            return;
        var emptyIndex = this.codeArr.indexOf(null);
        if (emptyIndex < this.maxLen - 1) {
            this.codeArr[emptyIndex] = num;
        }
        else if (emptyIndex === this.maxLen - 1) {
            this.codeArr[emptyIndex] = num;
            if (this.d.pinHandler) {
                this.d.pinHandler(this.getValues())
                    .then(function () {
                    _this.dismiss('done');
                })
                    .catch(function () {
                    _this.restoreClick();
                });
            }
            else {
                this.dismiss('done');
            }
        }
    };
    PincodeCmp.prototype.delClick = function () {
        var i = this.codeArr.length - 1;
        for (i; i > -1; i--) {
            if (this.codeArr[i] !== null) {
                this.codeArr[i] = null;
                break;
            }
        }
    };
    PincodeCmp.prototype.restoreClick = function () {
        this.codeArr = new Array(this.maxLen);
        this.codeArr.fill(null);
    };
    PincodeCmp.prototype.btnClick = function (button) {
        if (!this.enabled) {
            return;
        }
        this.lastClick = Date.now();
        var shouldDismiss = true;
        if (button.handler) {
            if (button.handler(this.getValues()) === false) {
                shouldDismiss = false;
            }
        }
        if (shouldDismiss) {
            this.dismiss(button.role).catch(function () {
                console.debug('alert can not be dismissed');
            });
        }
    };
    PincodeCmp.prototype.isNum = function (num) {
        if (util.isNumber(num)) {
            return num <= 9 || num >= 0;
        }
        else {
            return false;
        }
    };
    PincodeCmp.prototype.bdClick = function () {
        if (this.enabled && this.d.enableBackdropDismiss) {
            this.dismiss('backdrop');
        }
    };
    PincodeCmp.prototype.cancelClick = function () {
        this.dismiss('cancel');
    };
    PincodeCmp.prototype.forgotClick = function () {
        this.dismiss('forgot');
    };
    PincodeCmp.prototype.dismiss = function (role) {
        this.stateExpression = 'off';
        var opts = {
            minClickBlockDuration: 400
        };
        return this._viewCtrl.dismiss(this.getValues(), role, opts);
    };
    PincodeCmp.prototype.getValues = function () {
        return this.d.encoded(this.codeArr.join(''));
    };
    PincodeCmp.prototype.ngOnDestroy = function () {
        util.assert(this.gestureBlocker.blocked === false, 'gesture blocker must be already unblocked');
        this.gestureBlocker.destroy();
    };
    return PincodeCmp;
}());
PincodeCmp.decorators = [
    { type: core.Component, args: [{
                selector: 'ion-pincode',
                animations: [
                    core.trigger('openClose', [
                        core.state('off, void', core.style({ bottom: '-295px' })),
                        core.state('on', core.style({ bottom: '0' })),
                        core.transition('on <=> off', [core.animate("400ms cubic-bezier(.36,.66,.04,1)")])
                    ])
                ],
                template: "\n        <ion-backdrop (click)=\"bdClick()\" [class.backdrop-no-tappable]=\"!d.enableBackdropDismiss\"></ion-backdrop>\n        <div class=\"pincode-wrapper\"\n             [class.hide-toolbar]=\"d.hideToolbar\"\n             [@openClose]=\"stateExpression\">\n            <div class=\"pincode-toolbar\" *ngIf=\"!d.hideToolbar\">\n                <button *ngIf=\"!d.hideCancelButton\" ion-button clear (click)=\"cancelClick()\">{{d.cancelButtonText}}\n                </button>\n                <div class=\"pincode-title\">{{d.title}}</div>\n                <button *ngIf=\"!d.hideForgotPassword\" ion-button clear small class=\"r-btn\" (click)=\"forgotClick()\">\n                    {{d.forgotPasswordText}}\n                </button>\n            </div>\n            <div class=\"pincode-input\">\n                <ion-grid class=\"pincode-input-grid\">\n                    <ion-row class=\"pincode-input-row\">\n                        <ion-col *ngFor=\"let item of codeArr; let i = index\">\n                            <span *ngIf=\"!d.visibility || (!isNum(codeArr[i]))\" class=\"obscured\" [class.filling]=\"isNum(codeArr[i])\"></span>\n                            <span *ngIf=\"d.visibility\" class=\"code\">{{item}}</span>\n                        </ion-col>\n                    </ion-row>\n                </ion-grid>\n            </div>\n            <div class=\"pincode-button-wrapper\">\n                <ion-grid>\n                    <ion-row>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(1)\">1</button>\n                        </ion-col>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(2)\">2</button>\n                        </ion-col>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(3)\">3</button>\n                        </ion-col>\n                    </ion-row>\n                    <ion-row>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(4)\">4</button>\n                        </ion-col>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(5)\">5</button>\n                        </ion-col>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(6)\">6</button>\n                        </ion-col>\n                    </ion-row>\n                    <ion-row>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(7)\">7</button>\n                        </ion-col>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(8)\">8</button>\n                        </ion-col>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(9)\">9</button>\n                        </ion-col>\n                    </ion-row>\n                    <ion-row>\n                        <ion-col>\n                            <button ion-button color=\"light\" icon-only (click)=\"restoreClick()\">\n                                <ion-icon ios=\"ios-refresh\" name=\"md-refresh\"></ion-icon>\n                            </button>\n                        </ion-col>\n                        <ion-col>\n                            <button ion-button color=\"light\" (click)=\"numClick(0)\">0</button>\n                        </ion-col>\n                        <ion-col>\n                            <button ion-button color=\"light\" icon-only (click)=\"delClick()\">\n                                <ion-icon ios=\"ios-backspace\" name=\"md-backspace\"></ion-icon>\n                            </button>\n                        </ion-col>\n                    </ion-row>\n                </ion-grid>\n            </div>\n        </div>\n    ",
                styles: ["ion-pincode .pincode-wrapper{z-index:10;position:absolute;bottom:0;left:0;right:0;display:block;overflow:hidden;width:100%;max-width:500px;height:295px;margin:auto;background-color:#fcfcfc;border-top:1px solid #c8c7cc;contain:content}ion-pincode .pincode-wrapper.hide-toolbar{height:251px}ion-pincode .pincode-wrapper .button-md{height:4.6rem;font-size:1.8rem;color:#212121}ion-pincode .pincode-md .pincode-wrapper,ion-pincode .pincode-wp .pincode-wrapper{background-color:#fff}ion-pincode .pincode-toolbar{z-index:1;width:100%;contain:strict;display:block;height:44px;border-bottom:.55px solid #c8c7cc;background:#fff}ion-pincode .pincode-toolbar button{margin:0;height:44px;background:0 0;z-index:1}ion-pincode .pincode-toolbar .r-btn{float:right;text-align:left;font-size:12px;text-transform:none;font-weight:400}ion-pincode .pincode-toolbar .pincode-title{text-align:center;width:100%;line-height:44px;position:absolute;left:0;right:0;top:0;bottom:0}ion-pincode .pincode-md .pincode-toolbar,ion-pincode .pincode-wp .pincode-toolbar{border-bottom:none}ion-pincode .pincode-input{margin:16px}ion-pincode .pincode-input .pincode-input-grid{border:1px solid #e4e4e4}ion-pincode .pincode-input .pincode-input-row{text-align:center}ion-pincode .pincode-input .pincode-input-row ion-col{border-left:1px solid #e4e4e4;height:32px;line-height:32px;padding:0;text-align:center}ion-pincode .pincode-input .pincode-input-row ion-col:nth-child(1){border-left:0}ion-pincode .pincode-input .pincode-input-row ion-col span.obscured{width:1em;height:1em;border-radius:1em;display:inline-block;background-color:#eee}ion-pincode .pincode-input .pincode-input-row ion-col span.obscured.filling{background-color:#333}ion-pincode .pincode-input .pincode-input-row ion-col span.code{font-size:1.2em}ion-pincode .pincode-button-wrapper{position:absolute;width:100%;bottom:0}ion-pincode .pincode-button-wrapper .grid,ion-pincode .pincode-button-wrapper ion-col{padding:0}ion-pincode .pincode-button-wrapper ion-col button{border:none;margin:0;width:100%;background-color:#fff;color:#000;-webkit-box-shadow:none;box-shadow:none;border-radius:0}"],
                host: {
                    'role': 'dialog',
                    '[attr.aria-labelledby]': 'hdrId',
                    '[attr.aria-describedby]': 'descId'
                },
                encapsulation: core.ViewEncapsulation.None,
            },] },
];
PincodeCmp.ctorParameters = function () { return [
    { type: ionicAngular.ViewController, },
    { type: core.ElementRef, },
    { type: ionicAngular.Config, },
    { type: ionicAngular.GestureController, },
    { type: ionicAngular.NavParams, },
    { type: core.Renderer2, },
    { type: ionicAngular.Platform, },
]; };
var PinCode = /** @class */ (function (_super) {
    __extends(PinCode, _super);
    function PinCode(app, opts) {
        if (opts === void 0) { opts = {}; }
        var _this = this;
        opts.enableBackdropDismiss = util.isPresent(opts.enableBackdropDismiss) ?
            !!opts.enableBackdropDismiss : true;
        opts.hideForgotPassword = util.isPresent(opts.hideForgotPassword) ?
            !!opts.hideForgotPassword : false;
        opts.hideCancelButton = util.isPresent(opts.hideCancelButton) ?
            !!opts.hideCancelButton : false;
        opts.encoded = util.isFunction(opts.encoded) ? opts.encoded : function (c) {
            return c;
        };
        opts.title = opts.title || 'Password';
        opts.cancelButtonText = opts.cancelButtonText || 'cancel';
        opts.forgotPasswordText = opts.forgotPasswordText || 'forgot password';
        opts.passSize = opts.passSize || 6;
        _this = _super.call(this, PincodeCmp, opts, null) || this;
        _this._app = app;
        _this.isOverlay = true;
        return _this;
    }
    PinCode.prototype.getTransitionName = function (direction) {
        var key = (direction === 'back' ? 'alertLeave' : 'alertEnter');
        return this._nav && this._nav.config.get(key);
    };
    PinCode.prototype.setTitle = function (title) {
        this.data.title = title;
    };
    PinCode.prototype.setCssClass = function (cssClass) {
        this.data.cssClass = cssClass;
    };
    PinCode.prototype.present = function (navOptions) {
        if (navOptions === void 0) { navOptions = {}; }
        navOptions.minClickBlockDuration = navOptions.minClickBlockDuration || 400;
        return this._app.present(this, navOptions);
    };
    return PinCode;
}(ionicAngular.ViewController));
var PincodeController = /** @class */ (function () {
    function PincodeController(_app) {
        this._app = _app;
    }
    PincodeController.prototype.create = function (opts) {
        if (opts === void 0) { opts = {}; }
        return new PinCode(this._app, opts);
    };
    return PincodeController;
}());
PincodeController.decorators = [
    { type: core.Injectable },
];
PincodeController.ctorParameters = function () { return [
    { type: ionicAngular.App, },
]; };
var PincodeInputModule = /** @class */ (function () {
    function PincodeInputModule() {
    }
    return PincodeInputModule;
}());
PincodeInputModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [ionicAngular.IonicModule, common.CommonModule],
                declarations: [PincodeCmp],
                providers: [PincodeController],
                entryComponents: [PincodeCmp],
            },] },
];
PincodeInputModule.ctorParameters = function () { return []; };

exports.PinCode = PinCode;
exports.PincodeController = PincodeController;
exports.PincodeCmp = PincodeCmp;
exports.PincodeInputModule = PincodeInputModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
